### DISCUSSION TITLE
Is this considered to be an okay practice in nextjs middleware?

### DISCUSSION THREAD

**Discussion by josephp27** [supabase/supabase#20563] [Questions] [answered]
Hi, is this considered to be an ok practice? I am only doing this on my api routes and would like to reduce code overhead by authenticating and passing the user to the downstream api. Obviously, I don't want a security issue to happen. However, this afaik this is inject internally and would not be susceptible to a mitm attack

TL;DR If I inject this user object on the X-User header and restrict to only api/ endpoints, can a user somehow MITM and rewrite this header AFTER I wrote it

Thanks!

```js
import { createMiddlewareSupabaseClient } from "@supabase/auth-helpers-nextjs";
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const res = NextResponse.next();

  const supabase = createMiddlewareSupabaseClient({ req, res });

  const {
    data: { session },
  } = await supabase.auth.getSession();

  if (!session) {
    return new Response(null, { status: 401, statusText: "Not Authorized" });
  }

  // add the user to the request
  const requestHeaders = new Headers(req.headers);
  requestHeaders.set("X-User", JSON.stringify(session.user));

  const response = NextResponse.next({
    request: {
      headers: requestHeaders,
    },
  });

  return response;
}

// match anything that is an api
export const config = {
  matcher: "/api/:limits*",
};
```

**Comment by charislam** [comment, marked as answer]
@josephp27, what's the reason you want to avoid calling `supabase.auth.getUser()` from your API handlers? That seems like a more idiomatic solution

I haven't dug into how the middleware -> API handler flow works for Next, so I'm hesitant to answer that part of your question, but there is another security problem in your example code:

If you want to get the user from the server side (whether that's middleware or API handler), you should be using `supabase.auth.getUser()`, not `supabase.auth.getSession()`.

The reason is `getSession` isn't guaranteed to revalidate the Auth token with Supabase Auth server. It will only do so if the Auth token from the cookie is expired, otherwise it just returns the unencoded data from the cookie, which is very dangerous because anyone can spoof it. `getUser` will always go back to Supabase Auth server to revalidate so it is safe.

Before SSR got so popular, `getSession` was an OK pattern to use on the client, where nothing is trustworthy anyway, but it's an anti-pattern for the server. We've added a bunch of warnings to the docs recently about this.

**Comment by J0** [comment]
Hey @josephp27 Auth team here - would be keen to hear more about the rationale for avoiding `getUser()` as well if you're open to sharing. We're looking to improve the DX and a general idea of problems faced would be helpful

Lmk!
